library/alloc/src/vec/drain_filter.rs

```diff
@@ -0,0 +1,125 @@
+use core::ptr::{self};
+
+use super::{Vec, slice};
+
+/// An iterator which uses a closure to determine if an element should be removed.
+///
+/// This struct is created by [`Vec::drain_filter`].
+/// See its documentation for more.
+///
+/// # Example
+///
+/// ```
+/// #![feature(drain_filter)]
+///
+/// let mut v = vec![0, 1, 2];
+/// let iter: std::vec::DrainFilter<_, _> = v.drain_filter(|x| *x % 2 == 0);
+/// ```
+#[unstable(feature = "drain_filter", reason = "recently added", issue = "43244")]
+#[derive(Debug)]
+pub struct DrainFilter<'a, T, F>
+    where
+        F: FnMut(&mut T) -> bool,
+{
+    pub(crate) vec: &'a mut Vec<T>,
```

Comment:
@KodrAus said on zulip:

>  Hmm, I feel like PRs that are splitting up large modules into smaller child ones that use mostly `pub(super)` or `pub(crate)` internals could be missing an opportunity to define better privacy boundaries :thinking:

It'd be good to see if we can avoid making these all `pub(crate)`. `pub(super)` would already be a bit better, but it'd be good to check if we can make these private, possibly by adding a `pub(super)` function or two, depending on what the parent module actually needs.

If that's a lot of work, it would be okay to just merge this with `pub(super)` to avoid bitrot/merge conflicts, and fix it up in a smaller follow-up PR.